File: src/da4ml/cmvm/core/state_opr.py
from math import ceil, log2

import numpy as np
from numba import jit

from ..types import DAState, Op, QInterval
from ..util import csd_decompose


@jit
def _qint_add(qi1: QInterval, qi2: QInterval) -> QInterval:
    return QInterval(min=qi1.min + qi2.min, max=qi1.max + qi2.max, step=min(qi1.step, qi2.step))


@jit
def _qint_sub(qi1: QInterval, qi2: QInterval) -> QInterval:
    return QInterval(min=qi1.min - qi2.max, max=qi1.max - qi2.min, step=min(qi1.step, qi2.step))


@jit
def qint_add(qi1: QInterval, qi2: QInterval, sub0=False, sub1=False) -> QInterval:
    if sub0 != sub1:
        r = _qint_sub(qi1, qi2)
    else:
        r = _qint_add(qi1, qi2)
    if sub0:
        r = QInterval(min=-r.max, max=-r.min, step=r.step)
    return r


@jit
def cost_add(
    qint1: QInterval, qint2: QInterval, sub: bool = False, adder_size: int = -1, carry_size: int = -1
) -> tuple[float, float]:
    """Calculate the latency and cost of an addition operation.

    Parameters
    ----------
    qint1 : QInterval
        The first QInterval.
    qint2 : QInterval
        The second QInterval.
    sub : bool
        If True, the operation is a subtraction (a - b) instead of an addition (a + b).
    adder_size : int
        The atomic size of the adder.
    carry_size : int
        The size of the look-ahead carry.

    Returns
    -------
    tuple[float, float]
        The latency and cost of the addition operation.
    """
    if adder_size < 0 and carry_size < 0:
        return 1.0, 1.0
    if adder_size < 0:
        adder_size = 65535
    if carry_size < 0:
        carry_size = 65535

    f = -log2(min(qint1.step, qint2.step))
    min1, min2 = qint1.min, qint2.min
    max1, max2 = qint1.max, qint2.max
    if sub:
        min2, max2 = max2, min2
    max1, max2 = max1 + qint1.step, max2 + qint2.step
    i = ceil(log2(max(abs(min1), abs(min2), abs(max1), abs(max2))))
    k = int(qint1.min < 0 or qint2.min < 0)
    n_accum = k + i + f + 1
    # Align to the number of carry and adder bits, when they are block-based (e.g., 4/8 bits look-ahead carry in Xilinx FPGAs)
    # For Altera, the carry seems to be single bit adder chains, but need to check
    return float(ceil(n_accum / carry_size)), float(ceil(n_accum / adder_size))


@jit
def create_state(
    kernel: np.ndarray,
    qintervals: list[QInterval],
    inp_latencies: list[float],
    adder_size: int = -1,
    carry_size: int = -1,
    no_stat_init: bool = False,
):
    assert len(qintervals) == kernel.shape[0]
    assert len(inp_latencies) == kernel.shape[0]
    assert kernel.ndim == 2

    kernel = kernel.astype(np.float64)
    n_in, n_out = kernel.shape
    kernel = np.asarray(kernel)
    csd, shift0, shift1 = csd_decompose(kernel)
    n_bits = csd.shape[-1]
    expr = list(csd)
    shifts = (shift0, shift1)

    # Dirty numba typing trick
    stat = {Op(-1, -1, False, 0, 0.0, 0.0): 0}
    del stat[Op(-1, -1, False, 0, 0.0, 0.0)]
    expr_idx = list(range(len(expr)))

    # Loop over outputs, in0, in1, shift0, shift1 to gather all two-term pairs
    # Force i1>=i0
    if not no_stat_init:
        # Initialize the stat dictionary
        # Skip if no_stat_init is True (skip optimization)
        for i_out in range(n_out):
            for i0 in range(n_in):
                for j0 in range(n_bits):
                    bit0 = csd[i0, i_out, j0]
                    if not bit0:
                        continue
                    for i1 in range(i0, n_in):
                        for j1 in range(n_bits):
                            bit1 = csd[i1, i_out, j1]
                            if not bit1:
                                continue
                            # Avoid count the same bit
                            if i0 == i1 and j0 <= j1:
                                continue
                            dlat, dcost = cost_add(
                                qintervals[i0],
                                qintervals[i1],
                                bit0 != bit1,
                                adder_size=adder_size,
                                carry_size=carry_size,
                            )
                            pair = Op(i0, i1, bit0 != bit1, j1 - j0, dlat, dcost)
                            stat[pair] = stat.get(pair, 0) + 1

        for k in list(stat.keys()):
            if stat[k] < 2.0:
                del stat[k]

    ops = [Op(i, -1, False, 0, 0.0, 0.0) for i in range(n_in)]

    return DAState(
        shifts=shifts,
        expr_idx=expr_idx,
        expr=expr,
        ops=ops,
        latencies=inp_latencies,
        qintervals=qintervals,
        freq_stat=stat,
        kernel=kernel,
    )


@jit
def update_stats(
    state: DAState,
    op: Op,
    adder_size: int = -1,
    carry_size: int = -1,
):
    """Updates the statistics of any 2-term pair in the state that may be affected by implementing op."""
    id0, id1 = op.id0, op.id1

    ks = list(state.freq_stat.keys())
    for k in ks:
        if k.id0 == id0 or k.id1 == id1 or k.id1 == id0 or k.id0 == id1:
            del state.freq_stat[k]

    n_constructed = len(state.expr)
    modified = [n_constructed - 1]
    if id0 in state.expr_idx:
        modified.append(state.expr_idx.index(id0))
    if id1 != id0 and id1 in state.expr_idx:
        modified.append(state.expr_idx.index(id1))

    n_bits = state.expr[0].shape[-1]

    # Loop over outputs, in0, in1, shift0, shift1 to gather all two-term pairs
    for i_out in range(state.kernel.shape[1]):
        for _in0 in modified:
            for _in1 in range(n_constructed):
                if _in1 in modified and _in0 > _in1:
                    # Avoid double counting of the two locations when _i0 != _i1
                    continue
                # Order inputs, as _in0 can be either in0 or in1, range of _in is not restricted
                in0, in1 = (_in0, _in1) if _in0 <= _in1 else (_in1, _in0)
                for j0 in range(n_bits):
                    bit0 = state.expr[in0][i_out, j0]
                    if not bit0:
                        continue
                    for j1 in range(n_bits):
                        bit1 = state.expr[in1][i_out, j1]
                        if not bit1:
                            continue
                        if in0 == in1 and j0 <= j1:
                            continue
                        id0, id1 = state.expr_idx[in0], state.expr_idx[in1]
                        dlat, dcost = cost_add(
                            state.qintervals[in0],
                            state.qintervals[in1],
                            bit0 != bit1,
                            adder_size=adder_size,
                            carry_size=carry_size,
                        )
                        pair = Op(id0, id1, bit0 != bit1, j1 - j0, dlat, dcost)
                        state.freq_stat[pair] = state.freq_stat.get(pair, 0) + 1

    ks, vs = list(state.freq_stat.keys()), list(state.freq_stat.values())
    for k, v in zip(ks, vs):
        if v < 2.0:
            del state.freq_stat[k]
    return state


@jit
def gather_matching_idxs(state: DAState, op: Op):
    """Generates all i_out, j0, j1 ST expr[i_out][in0, j0] and expr[i_out][in1, j1] corresponds to op provided."""
    id0, id1 = op.id0, op.id1
    in0, in1 = state.expr_idx.index(id0), state.expr_idx.index(id1)
    shift = op.shift
    sub = op.sub
    n_out = state.kernel.shape[1]
    n_bits = state.expr[0].shape[-1]

    flip = False
    if shift < 0:
        in0, in1 = in1, in0
        shift = -shift
        flip = True

    sign = 1 if not sub else -1

    for j0 in range(n_bits - shift):
        for i_out in range(n_out):
            bit0 = state.expr[in0][i_out, j0]
            j1 = j0 + shift
            bit1 = state.expr[in1][i_out, j1]
            if sign * bit1 * bit0 != 1:
                continue

            if flip:
                yield i_out, j1, j0
            else:
                yield i_out, j0, j1


@jit
def update_expr(
    state: DAState,
    op: Op,
):
    "Updates the state by implementing the operation op, excepts common 2-term pair freq update."
    id0, id1 = op.id0, op.id1
    in0, in1 = state.expr_idx.index(id0), state.expr_idx.index(id1)
    sub = op.sub
    n_out = state.kernel.shape[1]
    n_bits = state.expr[0].shape[-1]

    expr_idx = state.expr_idx.copy()
    expr = state.expr.copy()
    latencies = state.latencies.copy()
    qintervals = state.qintervals.copy()
    ops = state.ops.copy()

    ops.append(op)

    new_slice = np.zeros((n_out, n_bits), dtype=np.int8)

    for i_out, j0, j1 in gather_matching_idxs(state, op):
        new_slice[i_out, j0] = expr[in0][i_out, j0]
        expr[in0][i_out, j0] = 0
        expr[in1][i_out, j1] = 0

    expr.append(new_slice)
    expr_idx.append(len(ops) - 1)
    qint0, qint1 = qintervals[in0], qintervals[in1]
    latencies.append(max(latencies[in0], latencies[in1]) + op.dlatency)
    qintervals.append(qint_add(qint0, qint1, sub1=sub))

    if in0 != in1:
        if in0 > in1:
            it = [in0, in1]
        else:
            it = [in1, in0]
    else:
        it = [in0]

    for i in it:
        if np.all(expr[i] == 0):
            expr.pop(i)
            expr_idx.pop(i)
            latencies.pop(i)
            qintervals.pop(i)

    return DAState(
        shifts=state.shifts,
        expr_idx=expr_idx,
        expr=expr,
        ops=ops,
        latencies=latencies,
        qintervals=qintervals,
        freq_stat=state.freq_stat,
        kernel=state.kernel,
    )


@jit
def update_state(
    state: DAState,
    pair_chosen: Op,
    adder_size: int,
    carry_size: int,
):
    """Update the state by removing all occurrences of pair_chosen from the state, register op code, and update the statistics."""
    state = update_expr(state, pair_chosen)
    state = update_stats(state, pair_chosen, adder_size=adder_size, carry_size=carry_size)
    return state

File: src/da4ml/_binary/cmvm/state_opr.cc
#include "state_opr.hh"
#include "bit_decompose.hh"
#include "src/da4ml/_binary/cmvm/types.hh"
#include <cmath>
#include <algorithm>
#include <unordered_map>
#include <array>

QInterval
qint_add(const QInterval &q0, const QInterval &q1, int64_t shift, bool sub0, bool sub1) {
    double min0 = q0.min, max0 = q0.max, step0 = q0.step;
    double min1 = q1.min, max1 = q1.max, step1 = q1.step;
    if (sub0) {
        std::swap(min0, max0);
        min0 = -min0;
        max0 = -max0;
    }
    if (sub1) {
        std::swap(min1, max1);
        min1 = -min1;
        max1 = -max1;
    }

    double s = std::pow(2.0, shift);
    min1 *= s;
    max1 *= s;
    step1 *= s;

    return QInterval{min0 + min1, max0 + max1, std::min(step0, step1)};
}

std::pair<double, double> cost_add(
    const QInterval &q0,
    const QInterval &q1,
    int64_t shift,
    bool sub,
    int adder_size,
    int carry_size
) {
    if (adder_size < 0 && carry_size < 0)
        return {1.0, 1.0};
    if (adder_size < 0)
        adder_size = 65535;
    if (carry_size < 0)
        carry_size = 65535;

    double min0 = q0.min, max0 = q0.max, step0 = q0.step;
    double min1 = q1.min, max1 = q1.max, step1 = q1.step;
    if (sub)
        std::swap(min1, max1);
    double sf = std::pow(2.0, shift);
    min1 *= sf;
    max1 *= sf;
    step1 *= sf;
    max0 += step0;
    max1 += step1;

    double f = -std::log2(std::max(step0, step1));
    double i = std::ceil(
        std::log2(
            std::max({std::abs(min0), std::abs(min1), std::abs(max0), std::abs(max1)})
        )
    );
    int k = (q0.min < 0 || q1.min < 0) ? 1 : 0;
    double n_accum = k + i + f;

    return {std::ceil(n_accum / carry_size), std::ceil(n_accum / adder_size)};
}

inline Pair _make_pair(int64_t id0, int64_t id1, int8_t v0, int8_t v1) {
    if (id0 > id1) {
        throw std::invalid_argument("id0 should be <= id1");
    }
    bool sub = SparseExpr::to_sign(v0) != SparseExpr::to_sign(v1);
    int8_t shift =
        static_cast<int8_t>(SparseExpr::to_shift(v1) - SparseExpr::to_shift(v0));
    return Pair{id0, id1, shift, sub};
}

DAState create_state(
    const xt::xarray<std::float32_t> &kernel_in,
    const std::vector<QInterval> &qintervals,
    const std::vector<double> &inp_latencies,
    bool no_stat_init
) {
    size_t n_in = kernel_in.shape(0);
    size_t n_out = kernel_in.shape(1);
    DAState state;

    xt::xarray<std::float32_t> kernel_f32(kernel_in);
    auto [csd, shift0, shift1] = csd_decompose(kernel_f32);

    // Zero out CSD rows for zero-range inputs
    for (size_t i = 0; i < n_in; ++i) {
        if (qintervals[i].min == 0.0 && qintervals[i].max == 0.0) {
            xt::view(csd, i) = 0;
        }
    }

    size_t n_bits = csd.shape(2);
    std::vector<SparseExpr> expr;
    for (size_t i = 0; i < n_in; ++i) {
        SparseExpr se;
        se.rows.resize(n_out);
        for (size_t io = 0; io < n_out; ++io) {
            for (size_t j = 0; j < n_bits; ++j) {
                int8_t v = csd(i, io, j);
                if (v != 0)
                    se.set_bit(io, j, v);
            }
        }
        expr.push_back(std::move(se));
    }

    auto &stat = state.freq_stat;
    if (!no_stat_init) {
        std::vector<Pair> raw_pairs;
        for (size_t i_out = 0; i_out < n_out; ++i_out) {
            for (size_t i0 = 0; i0 < n_in; ++i0) {
                const auto &row0 = expr[i0].rows[i_out];
                for (size_t i1 = i0; i1 < n_in; ++i1) {
                    const auto &row1 = expr[i1].rows[i_out];
                    if (row0.empty() || row1.empty())
                        continue;
                    if (i0 == i1) {
                        for (size_t a = 1; a < row0.size(); ++a) {
                            int8_t v0 = row0[a];
                            for (size_t b = 0; b < a; ++b) {
                                raw_pairs.push_back(_make_pair(i0, i1, v0, row0[b]));
                            }
                        }
                    }
                    else {
                        for (int8_t v0 : row0) {
                            for (int8_t v1 : row1) {
                                raw_pairs.push_back(_make_pair(i0, i1, v0, v1));
                            }
                        }
                    }
                }
            }
        }
        // Sort, deduplicate, count — only keep pairs with count >= 2
        stat.initialize(raw_pairs);
    }

    std::vector<Op> ops;
    for (size_t i = 0; i < n_in; ++i) {
        ops.push_back(Op{(int64_t)i, -1, -1, 0, qintervals[i], inp_latencies[i], 0.0});
    }

    state.shift0 = shift0;
    state.shift1 = shift1;
    state.expr = std::move(expr);
    state.n_bits = n_bits;
    state.ops = std::move(ops);
    // state.freq_stat = std::move(stat);
    state.kernel = kernel_in;
    return state;
}

std::vector<std::tuple<int64_t, int64_t, int64_t>>
gather_matching_idxs(const DAState &state, const Pair &pair) {
    int64_t id0 = pair.id0, id1 = pair.id1;
    int8_t shift = pair.shift;
    bool sub = pair.sub;
    size_t n_out = state.kernel.shape(1);

    bool flip = false;
    if (shift < 0) {
        std::swap(id0, id1);
        shift = -shift;
        flip = true;
    }

    int sign = sub ? -1 : 1;
    std::vector<std::tuple<int64_t, int64_t, int64_t>> result;

    for (size_t i_out = 0; i_out < n_out; ++i_out) {
        const auto &row0 = state.expr[id0].rows[i_out];
        const auto &row1 = state.expr[id1].rows[i_out];
        // Two-pointer merge on sorted rows
        size_t i0 = 0, i1 = 0;
        while (i0 < row0.size() && i1 < row1.size()) {
            int8_t shift0 = SparseExpr::to_shift(row0[i0]);
            int8_t j1 = SparseExpr::to_shift(row1[i1]);
            int8_t target_shift1 = shift0 + shift;
            if (j1 < target_shift1) {
                ++i1;
                continue;
            }
            if (j1 > target_shift1) {
                ++i0;
                continue;
            }
            // j1_val == target_j1
            int8_t bit0 = SparseExpr::to_sign(row0[i0]);
            int8_t bit1 = SparseExpr::to_sign(row1[i1]);
            if (sign * bit1 * bit0 == 1) {
                if (flip)
                    result.emplace_back(i_out, target_shift1, shift0);
                else
                    result.emplace_back(i_out, shift0, target_shift1);
            }
            ++i0;
            ++i1;
        }
    }
    return result;
}

Op pair_to_op(const Pair &pair, const DAState &state, int adder_size, int carry_size) {
    auto [dlat, cost] = cost_add(
        state.ops[pair.id0].qint,
        state.ops[pair.id1].qint,
        pair.shift,
        pair.sub,
        adder_size,
        carry_size
    );
    double lat =
        std::max(state.ops[pair.id0].latency, state.ops[pair.id1].latency) + dlat;
    QInterval qint = qint_add(
        state.ops[pair.id0].qint, state.ops[pair.id1].qint, pair.shift, false, pair.sub
    );
    return Op{pair.id0, pair.id1, (int64_t)pair.sub, pair.shift, qint, lat, cost};
}

void update_expr(DAState &state, const Pair &pair, int adder_size, int carry_size) {
    Op op = pair_to_op(pair, state, adder_size, carry_size);
    size_t n_out = state.kernel.shape(1);

    state.ops.push_back(op);

    SparseExpr new_slice;
    new_slice.rows.resize(n_out);

    int64_t id0 = pair.id0, id1 = pair.id1;
    int8_t rel_shift = pair.shift;
    bool sub = pair.sub;

    bool flip = false;
    if (rel_shift < 0) {
        std::swap(id0, id1);
        rel_shift = -rel_shift;
        flip = true;
    }

    int8_t target_sign = sub ? -1 : 1;

    for (size_t i_out = 0; i_out < n_out; ++i_out) {
        SparseExpr &expr0 = state.expr[id0];
        SparseExpr &expr1 = state.expr[id1];
        for (int8_t bit_loc0 = 0; bit_loc0 < (int8_t)expr0.rows[i_out].size();
             ++bit_loc0) {
            if (expr0.rows[i_out][bit_loc0] == 0)
                continue; // marked for removal
            auto [shift0, sign0] = expr0.pos_sign(i_out, bit_loc0);
            int8_t shift1 = shift0 + rel_shift;
            int8_t bit_loc1 = expr1.to_idx(i_out, shift1);
            if (shift1 >= state.n_bits)
                continue;
            int8_t sign1 =
                bit_loc1 >= 0 ? SparseExpr::to_sign(expr1.rows[i_out][bit_loc1]) : 0;
            if (target_sign * sign1 * sign0 != 1)
                continue;

            // Match found.
            if (!flip) {
                new_slice.set_bit(i_out, shift0, sign0);
            }
            else {
                new_slice.set_bit(i_out, shift1, sign1);
            }
            expr0.set_removal(i_out, bit_loc0);
            expr1.set_removal(i_out, bit_loc1);
        }

        expr0.compact(i_out);
        if (id0 != id1)
            expr1.compact(i_out);
    }

    state.expr.push_back(std::move(new_slice));
}

void update_stats(DAState &state, const Pair &pair) {
    int64_t id0 = pair.id0, id1 = pair.id1;

    // id0/1 are dirty now, purge all entries involving them.
    // Uses contiguous scan over FreqMap's flat vector instead of chasing
    // unordered_map node pointers.
    state.freq_stat.erase_if([&](const auto &e) {
        const Pair &p = e.first;
        return p.id0 == id0 || p.id0 == id1 || p.id1 == id0 || p.id1 == id1;
    });

    int64_t n_constructed = static_cast<int64_t>(state.expr.size());
    std::vector modified = {n_constructed - 1, id0};
    if (id0 != id1)
        modified.push_back(id1);

    size_t n_out = state.kernel.shape(1);

    // Collect all new pairs into a flat vector, then batch-merge.
    // This avoids per-pair hash lookups in the inner loops.
    std::vector<Pair> raw_pairs;

    for (size_t i_out = 0; i_out < n_out; ++i_out) {
        for (int64_t _in1 = 0; _in1 < n_constructed; ++_in1) {
            for (auto _in0 : modified) {
                if ((_in1 == n_constructed - 1 || _in1 == id0 || _in1 == id1) &&
                    _in0 > _in1)
                    continue;

                int64_t lo = std::min(_in0, _in1);
                int64_t hi = std::max(_in0, _in1);

                const auto &row_lo = state.expr[lo].rows[i_out];
                const auto &row_hi = state.expr[hi].rows[i_out];

                if (row_lo.empty() || row_hi.empty())
                    continue;

                if (lo == hi) {
                    const size_t sz = row_lo.size();
                    for (size_t a = 1; a < sz; ++a) {
                        for (size_t b = 0; b < a; ++b) {
                            raw_pairs.push_back(_make_pair(lo, lo, row_lo[a], row_lo[b]));
                        }
                    }
                }
                else {
                    for (int8_t v_hi : row_lo) {
                        for (int8_t v_lo : row_hi) {
                            raw_pairs.push_back(_make_pair(lo, hi, v_hi, v_lo));
                        }
                    }
                }
            }
        }
    }

    // Sort, deduplicate, count, and append — only pairs with count >= 2
    // are inserted. No merge needed since purge guarantees no overlap.
    state.freq_stat.batch_add(raw_pairs);
}

void update_state(DAState &state, const Pair &pair, int adder_size, int carry_size) {
    update_expr(state, pair, adder_size, carry_size);
    update_stats(state, pair);
}
